pkg/reconcile/reconcile.go

```diff
@@ -0,0 +1,129 @@
+
+// Operation describes the operation performed by CreateOrUpdate.
+type Operation string
+
+var (
+	Create   Operation = "create"
+	Recreate Operation = "recreate"
+	Update   Operation = "update"
+	None     Operation = "none"
+	Error    Operation = "error"
+)
+
+// ObjWithMeta describes a Kubernetes resource with a metadata field. It's a combination
+// of two common existing Kubernetes interfaces. We do this because we want to use methods
+// from each in CreateOrUpdate, whilst still keeping the argument type generic.
+type ObjWithMeta interface {
+	metav1.Object
+	runtime.Object
+}
+
+// CreateOrUpdate takes a Kubernetes object and a "diff function" and attempts to ensure
+// that the the object exists in the cluster with the correct state. It will use the diff
+// function to determine any differences between the cluster state and the local state and
+// use that to decide how to update it.
+func CreateOrUpdate(ctx context.Context, c client.Client, existing ObjWithMeta, kind string, diffFunc DiffFunc) (Operation, error) {
+	name := types.NamespacedName{
+		Namespace: existing.GetNamespace(),
+		Name:      existing.GetName(),
+	}
+	expected := existing.(runtime.Object).DeepCopyObject()
+	err := c.Get(ctx, name, existing)
+	if err != nil {
+		if !apierror.IsNotFound(err) {
+			return Error, err
+		}
+		if err := c.Create(ctx, existing); err != nil {
+			return Error, err
+		}
+		return Create, nil
+	}
+
+	// existing contains the state we just fetched from Kubernetes.
+	// expected contains the state we're trying to reconcile towards.
+	// If an update is required, DiffFunc will set the relevant fields on existing such that we
+	// can just resubmit it to the cluster to achieve our desired state.
+
+	op := diffFunc(expected, existing)
+	switch op {
+	case Recreate:
+		if err := c.Delete(ctx, existing); err != nil && !apierror.IsNotFound(err) {
+			return Error, err
+		}
+		if err := c.Create(ctx, expected); err != nil {
+			return Error, err
+		}
+		return Recreate, nil
+	case Update:
+		if err := c.Update(ctx, existing); err != nil {
+			return Error, err
+		}
+		return Update, nil
+	case None:
+		return None, nil
+	default:
+		return Error, fmt.Errorf("Unrecognised operation: %s", op)
+	}
+}
+
+// RoleDiff is a DiffFunc for Roles
+func RoleDiff(expectedObj runtime.Object, existingObj runtime.Object) Operation {
+	expected := expectedObj.(*rbacv1.Role)
+	existing := existingObj.(*rbacv1.Role)
+
+	if expected.ObjectMeta.Name != existing.ObjectMeta.Name {
+		return Recreate
+	}
+
+	if !reflect.DeepEqual(expected.Rules, existing.Rules) {
+		existing.Rules = expected.Rules
+		return Update
+	}
+
+	return None
+}
+
+// RoleBindingDiff is a DiffFunc for RoleBindings
+func RoleBindingDiff(expectedObj runtime.Object, existingObj runtime.Object) Operation {
+	expected := expectedObj.(*rbacv1.RoleBinding)
+	existing := existingObj.(*rbacv1.RoleBinding)
+	operation := None
+
+	if expected.ObjectMeta.Name != existing.ObjectMeta.Name {
```

Comment:
This is true, in our case. In future though, I can imagine we might have controllers managing resources where the name differs, and in that case we'd want to do this. I'll add a comment explaining that we don't currently make use of this but might want to - I'm somewhat hesitant to remove it in case we forget to account for it in the future. What do you think?

Reply:
I think we probably want to keep this, for a situation such as updating our code to produce slightly different resources: the names may change, so you'll want to recreate the resources.

This is thinking ahead of time but it's not too expensive for us to keep it now?